### PR TITLE
feat(electron): onboarding at first launch logic for client and web

### PR DESCRIPTION
**Pull request by CatsJuice**
- Added a simple abstraction of persistent storage class.
- Different persistence solutions are provided for web and client. 
    - web: stored in localStorage
    - client: stored in the application directory as `.json` file
- Define persistent app-config schema
- Add a new hook that can interactive with persistent-app-config reactively

### CODE REVIEW

**Comment by CatsJuice** [comment]
Current dependencies on/for this PR:
* `canary`
  * **PR #5183** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5183?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #5307** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5307?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5248** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5248?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5264** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5264?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #5283** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5283?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #5312** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5312?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #5319** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5319?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #5320** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5320?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #5322** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5322?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5216** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5216?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/5183?utm_source=stack-comment).

File: packages/frontend/electron/src/main/emitter.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import type { TypedEventEmitter } from '@toeverything/infra/core/event-emitter';
```
Comment (pengx17) [inline packages/frontend/electron/src/main/emitter.ts]: https://github.com/toeverything/AFFiNE/pull/5216/files#diff-5ee44df198617c4ac922685f1e6786e26713d887e7f1b03ba55a928bd7051d1dR7-R35

can you please check my changes in this closed PR? it no  longer uses event emitter and use dynamic import to get rid of circular dependency issue

Comment (CatsJuice) [reply packages/frontend/electron/src/main/emitter.ts]: Got it

File: packages/frontend/core/src/hooks/use-app-config-storage.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * @returns
+ */
+export function useAppConfigStorage(key?: keyof AppConfigSchema) {
+  const [_config, _setConfig] = useState(storage.get());
+
+  useEffect(() => {
+    storage.set(_config);
+  }, [_config]);
+
+  if (key) {
```
Comment (joooye34) [inline packages/frontend/core/src/hooks/use-app-config-storage.ts]: useMemo ?

File: packages/frontend/electron/src/main/onboarding.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import { join } from 'path';
+
+import { mainWindowOrigin } from './constants';
+// import { getExposedMeta } from './exposed';
+import { ensureHelperProcess } from './helper-process';
+import { logger } from './logger';
+
+// todo: not all window need all of the exposed meta
+const getWindowAdditionalArguments = async () => {
+  const { getExposedMeta } = await import('./exposed');
```
Comment (joooye34) [inline packages/frontend/electron/src/main/onboarding.ts:12]: why import in function?

Comment (CatsJuice) [reply packages/frontend/electron/src/main/onboarding.ts:12]: To avoid circular dependency, see https://github.com/toeverything/AFFiNE/pull/5251

**Review by joooye34 — APPROVED** [review]

**Pull request merged**